### PR TITLE
Truncate very long filter strings in tooltip

### LINKED ISSUE #44217: With large selection filters QGIS crashes when the mouse pointer hovers over the filter icon.
**Describe the bug**

Layer is based on a filter. The query filter contains a large query: "fid" IN (1,2,3,......) -> ..... represents hundreds or thousandsof ID's.
Whenever you hover the mouse on top of the layer name, layer name, source and a short representation of the query filter is shown in a yellow box.
Whenever you hover the mouse on top of the fil;ter icon, QGIS tries to show the complete query filter in the yellow popup-box. 
If the query is too large, i.e. contains hundreds or thousands of ID's, QGIS crashes.

**How to Reproduce**

Mouse pointer on top of the layer: Query is shortened. Selection contains about 5000 features.
afbeelding

Mouse pointer on top of the filter icon: QGIS tries to show the complete query.
afbeelding

If the query is larger, QGIS crashes.

**QGIS and OS versions**

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.18.3-Zürich | QGIS code revision | 735cc85be9
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.1.4 | Running against GDAL/OGR | 3.1.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | AutomaticBackup-master; BGTImport; clusterpoints; create_layer_from_selected_features; DataPlotly; debug_vs; GroupStats; itemmap_cropper; mmqgis; ORStools; pdokbaggeocoder; pdokservicesplugin; pdok_locatieserver_locator_filter; quick_map_services; TerrainShading; db_manager; processing

</body></html>

**Additional context**

Fastes way to create a selection layer with al large filter query: use the "create layer from selected features"-plugin

### PR DESCRIPTION
**Pull request by nyalldawson**
Fixes #44217

### CODE REVIEW

File: src/app/qgslayertreeviewfilterindicator.cpp [hunk 1]
```diff
@@ -49,15 +49,24 @@ QString QgsLayerTreeViewFilterIndicatorProvider::iconName( QgsMapLayer *layer )
...
-    return QStringLiteral( "<b>%1:</b><br>%2" ).arg( tr( "Filter" ), rlayer->subsetString().toHtmlEscaped() );
+    filter = rlayer->subsetString();
 
-  return QString();
+  if ( filter.isEmpty() )
+    return QString();
+
+  if ( filter.size() > 1024 )
+  {
+    filter = filter.left( 1023 ) + QString( QChar( 0x2026 ) );
```
Comment (nirvn) [inline src/app/qgslayertreeviewfilterindicator.cpp:66]: @nyalldawson , have you thought about truncating the middle part of the string instead? I.e. `FID IN (1, 2, 3, 4, ... 89, 90)` ?

**Review by nirvn — APPROVED** [review]

File: src/app/qgslayertreeviewfilterindicator.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/qgslayertreeviewfilterindicator.cpp]: Ok, done

Comment (nirvn) [reply src/app/qgslayertreeviewfilterindicator.cpp]: @nyalldawson , much appreciated. I love the utils function too, I assume it'll pop up in many places over the next few weeks/months :) )

**Comment by nirvn** [comment]
@nyalldawson , the Qt6 build test works:

```
  return string.leftRef( truncateFrom ) + QString( QChar( 0x2026 ) ) + string.midRef( truncateFrom + charactersToTruncate + 1 );
         ~~~~~~ ^
/usr/src/qgis/src/core/qgsstringutils.cpp:733:79: error: no member named 'midRef' in 'QString'
  return string.leftRef( truncateFrom ) + QString( QChar( 0x2026 ) ) + string.midRef( truncateFrom + charactersToTruncate + 1 );
                                                                       ~~~~~~ ^
2 errors generated.
```

:tada:

**Pull request merged**